### PR TITLE
Procedures

### PR DESCRIPTION
**Pull request by brennanjl**
This PR adds a procedural language to Kwil. It also attempts to simplify Kwil's dependency tree, both by removing the `parse` submodule (and its external dependencies), as well as unifying some unnecessarily duplicated types.

# Kuneiform Procedures

Kuneiform now supports a `procedure` block. This is very similar to the `action` block, with a few key changes:
1. The parameters are typed.
2. The returns are typed.
3. It supports basic arithmetic and control flow in its body.

These are, for all intents and purposes, just `actions`, but better. We still support actions, since all of our users use them, however we should expect they will migrate to procedures. The only tradeoff made by `procedures` is that they cannot use Kuneiform Precompiles.

# Parse

The `parse` submodule has been removed, and integrated into the main module. All of the grammars except for Kuneiform (sql, action, and now procedure) all live in `internal/parse`. Kuneiform can be found in the root directory.

The `sql` and `action` grammars were directly copied over from their old repos. `procedure` grammar is obviously totally new. `Kuneiform` was also _heavily_ changed. The old Kuneiform grammar had a ton of unused rules (I'd estimate ~50% of the grammar was not actually used), and its handling of action bodies was not suitable for procedure syntax. In addition to removing the unused grammar, there are a few key changes to Kuneiform:
1. It uses a separate lexer mode for parsing actions and procedures.
2. It uses recursive lexing for action/procedure bodies. Combined with the separate lexer mode, this allows us to put any arbitrary text inside of procedure/action bodies, and will only exit once enough right curly braces have been detected.
3. It now supports spaces for constraints and foreign key actions. E.g., to specify a primary key, you can use `primary`, `primary_key`, or `primary key`.

# engine/procedures

The bulk of the code for procedures can be found in `engine/procedures`. This includes:
1. Cleaning logic, providing case insensitivity.
2. Typing logic, which will type-check procedures before they are deployed. This is a significant improvement over pure PL/pgSQL, which does some type-checking, but often will not throw errors until runtime.
3. Generation of PL/pgSQL code.

# Schema Type

One of the most awkward parts of Kwil was how many different definitions we had of a parsed Kuneiform schema. There were definitions in Kuneiform, core/types/transactions, and common. All of these were used somewhat interchangeably, and had their own logic for verification, cleaning, etc.

This has now been unified into two definitions: `core/types` and `core/types/transactions`. They are very similar, however the definition in `core/types/transactions` is simpler, and meant to provide stable rlp encoding for transactions.

Everything should use the `core/types` definition, and the `transactions` definition should only be used when serializing to/from a transaction.
# TODO

There are a number of things that still need to be done before release. These should be done in subsequent PRs:

- Add SQL, action, and procedure syntax validation to Kuneiform. Previously, Kuneiform would perform this validation while parsing. Currently, the new version does not, however we should add it.
- `owner` does not currently work on procedures.
- non-`view` procedures can be called from within a `view` procedure. It will fail to mutate state, but a non-state-changing, non-`view` procedure can still be called.
- We still do not support uint256 and fixed point.

### CODE REVIEW

**Comment by brennanjl** [comment]
Ok it should be good for some review now. I still need to work on integration tests, but due overall I feel good about it.

There are still a few things missing from https://github.com/kwilteam/kwil-db/wiki/Kwil-v0.8, which will be added in subsequent PRs after this is merged:

- uint256 and fixed point types
- function(s) for foreign calling other schemas deployed

**Comment by brennanjl** [comment]
# Breaking Down The Review

Since this is a rather large PR, I've thought a bit about how this can be reviewed as easily as possible. I figured this could be split between Jon and Gavin, based on areas of focus:

Gavin should prioritize the parsing bits. This would be everything found in `kuneiform/` and `internal/parse/procedures`. He should also check out `internal/engine/sqlanalyzer/typing`, which is a package for determining resulting relations from SQL queries. I think this has a lot of overlap with query planning.

Jon should prioritize the typing and PL/pgSQL generation, found in `internal/engine/procedures`, `internal/engine/ddl`, and the minor changes in `internal/engine/execution`. This is more specific to the engine and Postgres.

Everyone should look at and give feedback to the refactoring of parse, as well as schema types. These are quite a bit smaller in terms of review work required, and are more a question of larger repo organization and design.

File: core/rpc/client/user/http/convert.go
Comment (jchappelow) [inline core/rpc/client/user/http/convert.go:308]: Need to `Returns: returns`?  The `returns` prepared above seems to be discarded.

File: core/types/schema.go
Comment (jchappelow) [inline core/types/schema.go:819]: It's required that only one (or neither) of Type or Table are set.  Maybe Clean should error if they both are?  Or leave that to other systems?  While RPC conversions helpers can catch it, is there logic else where in transaction execution that would?

File: core/types/schema.go
Comment (jchappelow) [inline core/types/schema.go:955]: Each of this string literals like `"blob"` are used three times in this file.   I think it's best to have a const block at the top defining them.

File: core/types/transactions/payloads.go
Comment (jchappelow) [inline core/types/transactions/payloads.go:278]: Just moved I know, but docs would be good here.  Not obvious what are types and the meaning of body (saved in event store and interpreted by some registered oracle, etc.).

File: core/types/validation/keywords.go
Comment (jchappelow) [inline core/types/validation/keywords.go:12]: Were these files supposed to be a `git mv` from `common/validation`?

File: internal/conv/conv.go
Comment (jchappelow) [inline internal/conv/conv.go:45]: I think this could use some docs.  the returned string isn't guaranteed to be valid UTF8 (won't be legible printed and will be rejected if attempting to use where a `TEXT` type is expected by postgresql).
What case does this support?  It's not obvious since this is in the `internal/conv` package so there could be many consumers with different needs.

Depending on the intent, maybe we want to employ https://pkg.go.dev/unicode/utf8#Valid before casting with no conversion.

**Review by jchappelow — COMMENTED** [review]
Just scraping the surface, but some comments to start.

File: core/client/client.go
Comment (brennanjl) [reply core/client/client.go:238]: Nope, this is a bug. Good catch.

File: core/types/schema.go
Comment (brennanjl) [reply core/types/schema.go:898]: Oh that's a good catch

File: internal/engine/procedures/generate/visitor.go
Comment (brennanjl) [reply internal/engine/procedures/generate/visitor.go:31]: Oh it's an extensions I use for generating methods to implement interfaces. I really don't know why it does that. Will fix.

File: core/types/validation/keywords.go
Comment (brennanjl) [reply core/types/validation/keywords.go:12]: Yes

File: internal/conv/conv.go
Comment (brennanjl) [reply internal/conv/conv.go:45]: Ah that's a good catch. The intent is to convert to a postgres TEXT type

File: internal/engine/execution/queries.go
Comment (jchappelow) [inline internal/engine/execution/queries.go:172]: I was wondering why we want a session variable instead of `SET LOCAL` so that there's no contamination or artifacts once the db transaction that wraps the procedure call commits/rollsback.

Also, can a procedure be called read-only / from RPC rather than in a consensus transaction?

File: internal/engine/procedures/procedural.go
Comment (jchappelow) [inline internal/engine/procedures/procedural.go:13]: `PreparedProcedures` => `GeneratePLPGSQL`

I also notice that you started with `procedural` as the package name and went with `procedures`, but there are some imports a docs like at the top of this file using `procedural`

File: internal/engine/procedures/traverse/traverse.go
Comment (jchappelow) [inline internal/engine/procedures/traverse/traverse.go:7]: Looks like a find/replace went wild here and on the `// package` line at the top.
Probably should say here `// Traverser is a procedure Visitor that traverses...`

File: internal/engine/procedures/typing/typing.go
Comment (jchappelow) [inline internal/engine/procedures/typing/typing.go:347]: I guess this is not going to be called normally, but it jumped out that it does `HasPrefix` twice in the positive path.  This would be more straightforward and efficient:

```go
func reverseCleanVar(v string) string {
	v, prefixed := strings.CutPrefix(v, "_param_")
	if !prefixed {
		panic("expected parameter prefix, received: " + v) // already in a panic, just do nothing and return v
	}
	return "$" + v
}
```

But if it's only ever going to be used inside another `panic`, might as well just return something and have the outer panic work using that regardless of if the prefix was there.

File: internal/engine/procedures/typing/typing.go
Comment (jchappelow) [inline internal/engine/procedures/typing/typing.go:473]: Is this a problem because we run the tree through the cleaner before typing?  I guess we can't reverse that order?
And the impact of the problem is that we just can't check that type, but it otherwise doesn't break anything?

File: internal/engine/procedures/typing/typing.go
Comment (jchappelow) [inline internal/engine/procedures/typing/typing.go:30]: And the user is prohibited by the antlr rules from declaring a variable starting with `_`?
Can we just return an error since there's no `recover` to catch this?

File: internal/engine/procedures/clean/clean.go
Comment (brennanjl) [reply internal/engine/procedures/clean/clean.go:189]: That's a good thing to be aware of, but I don't think it is a concern here. Everything that gets passed here is from the parser, and therefore won't even get parsed if it does not:
- start with a $ or @
- have a letter directly after
- have only letters, numbers, or underscores after that.

Comment (brennanjl) [reply internal/engine/procedures/clean/clean.go:189]: This is defined in the `VARIABLE` lexer rules in `internal/parse/procedure/grammar/ProcedureLexer.g4`.

File: internal/engine/procedures/clean/clean.go
Comment (brennanjl) [reply internal/engine/procedures/clean/clean.go:192]: Yep, this is just to avoid collisions since we can't use `@` and`$` in plpgsql variables, and plpgsql is very pick with naming collisions (e.g. variables and columns cannot collide).

File: internal/engine/procedures/clean/clean.go
Comment (brennanjl) [reply internal/engine/procedures/clean/clean.go:249]: Correct, this is impossible. INSERT/UPDATE/DELETE must be defined as top-level statements.Our syntax does not support nesting them in queries (I'm not even sure if any SQL dialect supports this, but we certainly do not)

File: internal/engine/execution/queries.go
Comment (brennanjl) [reply internal/engine/execution/queries.go:172]: > I was wondering why we want a session variable instead of SET LOCAL so that there's no contamination or artifacts once the db transaction that wraps the procedure call commits/rollsback.

I did not know SET LOCAL was a thing. Good call.

> Also, can a procedure be called read-only / from RPC rather than in a consensus transaction?

Yes. It should set the caller if one is included (which is only enforced if kgw is used), and the txid will be set to a blank string.

File: internal/engine/procedures/typing/typing.go
Comment (brennanjl) [reply internal/engine/procedures/typing/typing.go:473]: Correct. Still working on a workaround for this.

**Comment by brennanjl** [comment]
fyi @jchappelow I just pushed some upgrades to the engine schema. I figured this should go in this PR b/c it is needed for foreign calls to other schemas, which I'll be pushing soon.

I planned to include this in a future pr, but decided to include it here b/c it's not very hard to add but is a pretty important change.

File: internal/engine/procedures/clean/clean.go
Comment (jchappelow) [reply internal/engine/procedures/clean/clean.go:199]: > * have a letter directly after
> 
> * have only letters, numbers, or underscores after that.

OK, that's certainly constraining it to ASCII.  No issue here then.

File: internal/engine/execution/queries.go
Comment (brennanjl) [reply internal/engine/execution/queries.go:71]: Oh yeah good catch.

File: internal/engine/procedures/clean/clean.go
Comment (jchappelow) [reply internal/engine/procedures/clean/clean.go:273]: OK sounds good.

But for sake of clarity, I'm describing something like the following, which is straightforward for postgres:

```
# CREATE TABLE books (
    book_id SERIAL PRIMARY KEY,
    title VARCHAR(255),
    author VARCHAR(100),
    is_sold BOOLEAN DEFAULT FALSE
);


# INSERT INTO books (title, author, is_sold) VALUES
('The Great Gatsby', 'F. Scott Fitzgerald', FALSE),
('1984', 'George Orwell', FALSE),
('To Kill a Mockingbird', 'Harper Lee', FALSE),
('Moby Dick', 'Herman Melville', TRUE);  -- This book is already sold

INSERT 0 4

# WITH removed_books AS (
    DELETE FROM books
    WHERE is_sold = FALSE
    RETURNING book_id, title, author
)
SELECT * FROM removed_books;
┌─────────┬───────────────────────┬─────────────────────┐
│ book_id │         title         │       author        │
├─────────┼───────────────────────┼─────────────────────┤
│       1 │ The Great Gatsby      │ F. Scott Fitzgerald │
│       2 │ 1984                  │ George Orwell       │
│       3 │ To Kill a Mockingbird │ Harper Lee          │
└─────────┴───────────────────────┴─────────────────────┘
(3 rows)

Time: 14.023 ms
```

Or with the returning from an update:

```
psql://postgres@[local]:5432/kwil_test_db 
# WITH sold_books AS (
    UPDATE books
    SET is_sold = TRUE
    WHERE author = 'Harper Lee' AND is_sold = FALSE
    RETURNING book_id, title, author
)
SELECT * FROM sold_books;
┌─────────┬───────────────────────┬────────────┐
│ book_id │         title         │   author   │
├─────────┼───────────────────────┼────────────┤
│       3 │ To Kill a Mockingbird │ Harper Lee │
└─────────┴───────────────────────┴────────────┘
(1 row)

Time: 12.565 ms
```

This is quite contrived, but I wanted to be sure it didn't slip past this code.

**Review by Yaiba — COMMENTED** [review]
This is the first batch of my review.  
Just started to review sqlanalyzer.typing, it will takes more times from here so I just submit those reviews first.

File: internal/engine/sqlanalyzer/clean/walker.go
Comment (Yaiba) [reply internal/engine/sqlanalyzer/clean/walker.go:92]: ah ignore this, I delete it in IDE but apparently it didn't work.

File: internal/parse/action/ast_builder.go
Comment (Yaiba) [inline internal/parse/action/ast_builder.go:340]: Yeah, it's coupled to what engine defines. We'll need to declare what's supported in this pkg, it's only a list of names

File: internal/parse/sql/tree/expression.go
Comment (Yaiba) [inline internal/parse/sql/tree/expression.go:203]: Why this check is removed?

File: kuneiform/parser.go
Comment (Yaiba) [inline kuneiform/parser.go:1]: We'll also need wasm here

File: internal/parse/action/ast_builder.go
Comment (Yaiba) [reply internal/parse/action/ast_builder.go:340]: hmm, wait, we used to have a function list in `tree` package

File: internal/engine/sqlanalyzer/clean/walker.go
Comment (brennanjl) [reply internal/engine/sqlanalyzer/clean/walker.go:92]: correct

File: internal/engine/sqlanalyzer/typing/typing.go
Comment (brennanjl) [reply internal/engine/sqlanalyzer/typing/typing.go:137]: II can't really. I'm not actually sure if it is possible, but I haven't spent much time thinking through the edge cases.

File: core/types/schema.go
Comment (brennanjl) [reply core/types/schema.go:32]: Yes, it tests name uniqueness

File: core/types/validation/name.go
Comment (brennanjl) [reply core/types/validation/name.go:55]: Code is actually not used, deleting

File: core/types/validation/restrictions.go
Comment (brennanjl) [reply core/types/validation/restrictions.go:5]: LOL this is code from before you joined. I am shocked it has made it this far. This code itself has not been used for a very long time, will delete.

Comment (brennanjl) [reply core/types/validation/restrictions.go:5]: deleting

File: core/types/validation/restrictions.go
Comment (brennanjl) [reply core/types/validation/restrictions.go:6]: All of this code is simply moved over. I have not changed any of it. I change the ident length to 32 in a new PR though

File: core/types/schema.go
Comment (Yaiba) [reply core/types/schema.go:32]: I mean why we started to require uniqueness?  I can make sense of action/procedure global uniqueness. 
But if this is just the behavior we want to change, it's fine

File: internal/engine/sqlanalyzer/typing/visitor_test.go
Comment (brennanjl) [reply internal/engine/sqlanalyzer/typing/visitor_test.go:144]: wow good catch

File: internal/engine/sqlanalyzer/typing/visitor.go
Comment (brennanjl) [reply internal/engine/sqlanalyzer/typing/visitor.go:480]: Ah yeah this is another really good catch

File: internal/engine/sqlanalyzer/typing/visitor.go
Comment (brennanjl) [reply internal/engine/sqlanalyzer/typing/visitor.go:656]: Is there a problem here? I cannot see it

File: core/types/schema.go
Comment (brennanjl) [reply core/types/schema.go:934]: No, most of the time we care if a type is null I think.

I have been thinking about getting rid of null type though, since it isn't really a type.

File: internal/parse/sql/grammar/SQLParser.g4
Comment (brennanjl) [reply internal/parse/sql/grammar/SQLParser.g4:333]: I really don't know. I didn't change the SQL parser, it is just copied over from the repo where you made it 😆

File: internal/parse/action/ast_builder.go
Comment (brennanjl) [reply internal/parse/action/ast_builder.go:340]: Yeah I moved it

Comment (brennanjl) [reply internal/parse/action/ast_builder.go:340]: Most of it was unused and it didn't include type info

File: internal/parse/sql/tree/expression.go
Comment (brennanjl) [reply internal/parse/sql/tree/expression.go:203]: We now remove it

File: internal/engine/sqlanalyzer/typing/visitor.go
Comment (Yaiba) [reply internal/engine/sqlanalyzer/typing/visitor.go:658]: not a problem, just a bit confusing, since `Equals` should just tell if two type are same. That's why I suggested changes in `Equals`

File: internal/parse/sql/grammar/SQLParser.g4
Comment (Yaiba) [reply internal/parse/sql/grammar/SQLParser.g4:333]: I get it. And I think we should change this name.

File: internal/engine/sqlanalyzer/typing/visitor.go
Comment (Yaiba) [reply internal/engine/sqlanalyzer/typing/visitor.go:658]: If one of the type is NULL  and another is not NULL, `Equals` should return false

**Review by Yaiba — APPROVED** [review]
I think this pr is good to go

**Review by jchappelow — APPROVED** [review]

**Pull request merged**
